addons/sale_stock/models/sale_order.py

```diff
@@ -142,10 +142,25 @@ def action_view_delivery(self):
     def action_cancel(self):
         documents = None
         for sale_order in self:
+            if any(picking.state == 'done' for picking in sale_order.picking_ids) and not self._context.get('disable_cancel_warning'):
+                note = 'Are you sure to cancel this sales order? Some products have already been delivered.'
+                if sale_order.invoice_ids.filtered(lambda inv: inv.state == 'draft'):
+                    note += '<br/>The order has been already invoiced in draft. The invoice will be cancelled too.'
+                return {
+                    'name': _('Cancel Sales Order'),
+                    'view_mode': 'form',
+                    'res_model': 'sale.order.cancel',
+                    'view_id': self.env.ref('sale_stock.sale_order_cancel_view_form').id,
+                    'type': 'ir.actions.act_window',
+                    'context': {'default_sale_id': self.id, 'default_note': note},
```

Comment:
Please adapt the code to make the note translatable ;).

Reply:
@Feyensv  Done.